### DISCUSSION TITLE
Upgrade Jedis from version 3 to 5

### DISCUSSION THREAD

**Discussion by tamasgyenese** [xetorthio/jedis#3947] [Q&A] [answered]
Hi all!
I'm on the way to upgrade Jedis from version 3.x to 5.y and I realized the Connection class was made as the base of operations, instead of Jedis class. What is the benefit of that? Is it recommended to use Connection instead of Jedis?
And I also noticed that the BinaryJedis class has been removed, so it means the JedisClusterConnectionHandler has also been removed. What is the new alternative for that, ClusterConnectionProvider class? Same question for the JedisClusterCommand class.

I'm sorry if the answers are obvious, just I don't which would be the preferred way.
Thank you!

**Comment by sazzad16** [comment, marked as answer]
@tamasgyenese Thank you for using Jedis.

It has been close to 3 years Jedis 4.0.0 has been released. Almost none of the people that were involved in Jedis are now with us. We know what we have but it is now difficult to go through all the reasoning.

Still, to mention something:
- Connection was base in many cases even in 3.x but due to the design we always had to create the Jedis object. Now we can avoid that.
- Connection allows us different new designs. For example, in #3757, currently we have a new CacheConnection class which is actually a sub-class of Connection.

For rest of the questions, I would suggest you to provide code examples for 3.x so we may be help to update those.

**Reply by tamasgyenese** [reply]
Hi @sazzad16 !
Thank you for your response. I started walk through the implementation and now lot of things make sense. I would just one more question (at least for now), if I saw well, within the DefaultJedisClientConfig class there is a readOnlyForRedisClusterReplicas property, so, If i would like read only from replicas, should I create a JedisCluster with that config and using executeCommandToReplica method? I would like to execute readonly scripts only on replicas.
Thank you for your work and time!

**Reply by sazzad16** [reply]
@tamasgyenese Yes, you can create a JedisCluster with that config AND use executeCommandToReplica method.

BTW, ATM that config and method is only available in 5.2.0-SNAPSHOT and 5.2.0-beta5.

**Comment by tamasgyenese** [comment]
@sazzad16  thank you! 
And please could you help me with that one:
Currently I extend the `JedisClusterCommand<T>` class and implement `public T execute(Jedis j)` and calling the `run()`, something like this:
```java
public class MyClass<T> extends JedisClusterCommand<T> {
    private final Function<WrapperClass, T> function;

    public MyClass(JedisClusterConnectionHandler connectionHandler, Function<WrapperClass, T> command, int maxAttempts) {
        super(connectionHandler, maxAttempts, Duration.ofMillis(30));
        function = command;
    }

    @Override
    public T execute(Jedis jedis) {
        return function.apply(new WrapperClass(jedis));
    }
}
```
and calling it:
```java
return new MyClass<T>(connectionHandler, command, 4).run(key);
```
I've checked the CommandExecutor classes, like ClusterCommandExecutor, but the methods are `final`, or I missed something. So, I'm looking for something similar, but with using Connection and my extended version of ClusterConnectionProvider.
thank you for your advice!

**Reply by sazzad16** [reply]
@tamasgyenese You can't use these directly from JediCluster, right? In that case, you won't be to do that in updated versions.

`JedisClusterCommand` and etc were public classes but for internal purposes. So these are not kept when the library was restructured.

**Reply by tamasgyenese** [reply]
@sazzad16 
Thank you so much all of your answers!

**Reply by tamasgyenese** [reply]
Hi @sazzad16 

Please could you help me with the following?
I'm planning to use `CommandExecutor::executeCommand` and it can be `ClusterCommandExecutor` or `SimpleCommandExecutor` based on the Redis type. I really like the `ClusterCommandExecutor`, because the retryable process is very easy to configure thanks to the provider. But it's possible to execute a `SLOWLOG GET` command with them? I've seen an example in `Jedis::slowlogGet`, but i'm not sure how I can do it.

Thanks,
Tamas

**Reply by sazzad16** [reply]
If you execute a `SLOWLOG GET` command using ClusterCommandExecutor, then it'll be executed on a random node in the cluster. Is this what you're looking for?

**Reply by tamasgyenese** [reply]
Thank you for your answer, after reading it twice, maybe it makes no sense, except it returns from all the nodes :).
So, what if I would like to execute the `SLOWLOG GET` and/or `CLUSTER SLOTS` commands with using a `SimpleCommandExecutor` (or any other class which implements the `CommandExecutor` interface), is it possible or to do that, I would need a `Jedis` instance? 
Thanks,
Tamas

**Reply by sazzad16** [reply]
@tamasgyenese Please ask what you actually need. Do you need to execute a command (i.e. SLOWLOG GET) to all cluster nodes?

**Reply by tamasgyenese** [reply]
Sure, I'm sorry. I would like to execute commands like `SLOWLOG GET/ CLUSTER SLOTS` using `SimpleCommandExecutor ` without creating `Jedis` instance.

**Reply by sazzad16** [reply]
Just create proper CommandObject and send it to the executor.

SLOWLOG GET:

```java
List<Object> rawSlowlog = executor.executeCommand(new CommandObject(
        new CommandArguments(Protocol.Command.SLOWLOG).add(Protocol.Keyword.GET),
        BuilderFactory.RAW_OBJECT_LIST));
List<Slowlog> slowlog = Slowlog.from(rawSlowlog);
```

CLUSTER SLOTS:

```java
executor.executeCommand(new CommandObject(
        new CommandArguments(Protocol.Command.CLUSTER).add(Protocol.Keyword.SLOTS),
        BuilderFactory.RAW_OBJECT_LIST));
```

**Reply by tamasgyenese** [reply]
Hi @sazzad16 
Thank you so much your time and all of your answers!

**Reply by sazzad16** [reply]
@tamasgyenese You're welcome!

**Reply by tamasgyenese** [reply]
Hi @sazzad16 

out of curiosity, what is the reason that the `UnifiedJedis` does not implement `Commands` as `Jedis` does, e.g. `ClusterCommands` ?

**Reply by sazzad16** [reply]
@tamasgyenese We didn't have any directive about connection specific and/or node specific commands from a _generic_ object like UnifiedJedis. By generic I mean it could be anything; i.e. standalone server, sentinels, cluster, etc.
